dimod/higherorder/utils.py

```diff
@@ -174,6 +175,94 @@ def reduce_binary_polynomial(poly: BinaryPolynomial) -> Tuple[
+    if vartype is None:
+        if qm is None:
+            raise ValueError("one of vartype or qm must be provided")
+        else:
+            vartype = qm.vartype
+    else:
+        vartype = as_vartype(vartype)  # handle other vartype inputs
+        if qm is None:
+            qm = qm_factory(vartype)
+        else:
+            qm = qm.change_vartype(vartype, inplace=False)
+
+    # for backwards compatibility, add an info field
+    if not hasattr(qm, 'info'):
+        qm.info = {}
+    qm.info['reduction'] = {}
+    return qm, vartype
+
+def _init_binary_polynomial(poly, vartype):
+    if not (isinstance(poly, BinaryPolynomial) and (vartype in (poly.vartype, None))):
+        poly = BinaryPolynomial(poly, vartype=vartype)
+    return poly
+
+def _init_objective(bqm, reduced_terms):
+    for term, bias in reduced_terms:
+        if len(term) == 2:
+            bqm.add_interaction(*term , bias)
+        elif len(term) == 1:
+            bqm.add_variable(*term, bias)
+        elif len(term) == 0:
+            bqm.offset += bias
+        else:
+            # still has higher order terms, this shouldn't happen
+            msg = ('Internal error: not all higher-order terms were reduced. '
+                   'Please file a bug report.')
+            raise RuntimeError(msg)
+
+def make_quadratic_cqm(poly, vartype=None, cqm=None):
+    """Create a constrained quadratic model from a higher order polynomial.
+
+    Args:
+        poly (dict):
+            Polynomial as a dict of form {term: bias, ...}, where `term` is a tuple of
+            variables and `bias` the associated bias.
+
+        vartype (:class:`.Vartype`/str/set, optional):
+            Variable type for the binary quadratic model. Accepted input values:
+
+            * :class:`.Vartype.SPIN`, ``'SPIN'``, ``{-1, 1}``
+            * :class:`.Vartype.BINARY`, ``'BINARY'``, ``{0, 1}``
+
+            If `poly is a BinaryPolynomial , `vartype` is not required.
+
+        cqm (:class:`.BinaryQuadraticModel`, optional):
+            The terms of the reduced polynomial are added to this constrained quadratic model.
+            If not provided, a new constrained quadratic model is created.
+
+    Returns:
+        :class:`.ConstrainedQuadraticModel`
+
+    Examples:
+
+        >>> poly = {(0,): -1, (1,): 1, (2,): 1.5, (0, 1): -1, (0, 1, 2): -2}
+        >>> cqm = dimod.make_quadratic_cqm(poly, dimod.SPIN)
+
+    """
+    if not (vartype or isinstance(poly, BinaryPolynomial)):
+        raise ValueError("can not infer vartype")
+    cqm = cqm or ConstrainedQuadraticModel()
+    vartype = vartype or poly.vartype
+    poly = _init_binary_polynomial(poly, vartype)
+    variables = set().union(*poly)
+    reduced_terms, constraints = reduce_binary_polynomial(poly)
+
+    def var(x):
+        return BinaryQuadraticModel({x: 1.0}, {}, 0.0, vartype)
+
+    for (u, v), p in constraints:
+        cqm.add_constraint( var(u)*var(v) - var(p)  == 0, label = f'var__{u}*var_{v} == var_{p}')
```

Comment:
why not just `label=f'{u!r}*{v!r}=={p!r}`? might be confusing to later look for `var__{u}` in the `.variables`.

Reply:
> 
> 
> why not just `label=f'{u!r}*{v!r}=={p!r}`? might be confusing to later look for `var__{u}` in the `.variables`.

I thought about this. Main reaso for the `var_` is that the equation will look weird if the variables are numbers. Maybe that is ok. or maybe use something like `v{u}` or `x{u}`.. Not sure. I do not have strong feelings about this.

Reply:
Ok, in that case I think I am comfortable with the weird integer case. It is just a label after all and should be pretty clear from context.

Reply:
maybe wrap quotes?